nixos/modules/virtualisation/xen-dom0.nix

```diff
@@ -157,354 +285,594 @@ in
+                  '';
+                };
+                port = lib.options.mkOption {
+                  default = "/proc/xen/xsd_port";
+                  example = cfg.store.settings.xenstored.xenfs.port;
+                  type = lib.types.path;
+                  visible = false;
+                  description = ''
+                    Path to the Xen Store Daemon userspace port inside the XenFS pseudo-filesystem.
+                    While it is possible to alter this value, some drivers may be hardcoded to follow the default paths.
+                  '';
+                };
+              };
+            };
+            ringScanInterval = lib.options.mkOption {
+              default = 20;
+              example = 30;
+              type = lib.types.addCheck (
+                lib.types.int
+                // {
+                  name = "nonzeroInt";
+                  description = "nonzero signed integer, meaning !=0";
+                  descriptionClass = "nonRestrictiveClause";
+                }
+              ) (n: n != 0);
+              description = ''
+                Perodic scanning for all the rings as a safenet for lazy clients.
+                Define the interval in seconds; set to a negative integer to disable.
+              '';
+            };
+          };
+        };
+      };
+    };
   };
 
-  ###### implementation
+  ## Implementation ##
 
-  config = mkIf cfg.enable {
+  config = lib.modules.mkIf cfg.enable {
     assertions = [
       {
         assertion = pkgs.stdenv.isx86_64;
-        message = "Xen currently not supported on ${pkgs.stdenv.hostPlatform.system}";
+        message = "Xen is currently not supported on ${pkgs.stdenv.hostPlatform.system}.";
       }
       {
-        assertion = config.boot.loader.grub.enable && (config.boot.loader.grub.efiSupport == false);
-        message = "Xen currently does not support EFI boot";
+        assertion =
+          config.boot.loader.systemd-boot.enable
+          || (config.boot ? lanzaboote) && config.boot.lanzaboote.enable;
+        message = "Xen only supports booting on systemd-boot or Lanzaboote.";
+      }
+      {
+        assertion = cfg.dom0Resources.maxMemory >= cfg.dom0Resources.memory;
+        message = ''
+          You have allocated more memory to dom0 than virtualisation.xen.dom0Resources.maxMemory
+          allows for. Please increase the maximum memory limit, or decrease the default memory allocation.
+        '';
+      }
+      {
+        assertion = cfg.debug -> cfg.trace;
+        message = "Xen's debugging features are enabled, but logging is disabled. This is most likely not what you want.";
+      }
+      {
+        assertion = cfg.store.settings.quota.maxWatchEvents >= cfg.store.settings.quota.maxOutstanding;
+        message = ''
+          Upstream Xen recommends that maxWatchEvents be equal to or greater than maxOutstanding,
+          in order to mitigate denial of service attacks from malicious frontends.
+        '';
```

Comment:
Are these sufficiently bad configurations that we want them to be assertions, rather than warnings? (I’m perfectly fine with ruling out footguns like this; just want to check.)

Reply:
The quota stuff should certainly be kept, otherwise users might unknowingly open themselves to a known security vulnerability. The tracing stuff is more of a footgun, but could be replaced with a `lib.warn` with no issues. I don't see why someone would want to debug Xen without logs though.